### PR TITLE
(graph): replace cycle checking logic from `DFS` to ancestor set

### PR DESCRIPTION
**Pull request by Ocean-Ch**

### CODE REVIEW

File: internal/graph/dag.go [hunk 1]
```diff
@@ -39,9 +41,10 @@ func (g *DAG[VertexT]) AddEdge(from, to VertexT) error {
 		return ErrAcyclicConstraintViolation
```
Comment (Ocean-Ch) [inline internal/graph/dag.go:41]: is it intended that adding duplicate edges returns normally?

Comment (skjdfhkskjds) [reply internal/graph/dag.go:41]: yeah that should be fixed

Comment (Ocean-Ch) [reply internal/graph/dag.go:41]: ok yeah

**Comment by Ocean-Ch** [comment]
https://github.com/skjdfhkskjds/depinject/pull/10

**Pull request closed without merging**